python/ray/experimental/gpu_object_manager/gpu_object_manager.py

```diff
@@ -58,27 +35,41 @@ def __init__(self):
         # avoid circular import and because it imports third-party dependencies
         # like PyTorch.
         self._gpu_object_store: Optional["GPUObjectStore"] = None
+        # Lock to synchronize access to the GPU object store. This is needed
+        # because the GPU object store is accessed by both user threads, to
+        # put/get objects into the object store from user tasks, and by a
+        # background system thread, which handles data transfers and GC.
+        self.gpu_object_store_lock = threading.RLock()
 
     @property
     def gpu_object_store(self) -> "ray.experimental.GPUObjectStore":
-        if self._gpu_object_store is None:
-            from ray.experimental.gpu_object_manager.gpu_object_store import (
-                GPUObjectStore,
-            )
+        with self.gpu_object_store_lock:
+            if self._gpu_object_store is None:
+                from ray.experimental.gpu_object_manager.gpu_object_store import (
+                    GPUObjectStore,
+                )
 
-            self._gpu_object_store = GPUObjectStore()
+                self._gpu_object_store = GPUObjectStore(self.gpu_object_store_lock)
```

Comment:
Sharing the lock across these two classes is scary. From what I can tell, the lock is only used in `GPUObjectManager` to guard the lazy initialization of the `GPUObjectStore`. I suggest introducing a separate lock only for that purpose, then keeping the re-entrant lock and condition variable isolated as implementation details of the `GPUObjectStore`.